corehq/form_processor/reprocess.py

```diff
@@ -0,0 +1,146 @@
+from corehq.form_processor.interfaces.processor import FormProcessorInterface
+from corehq.form_processor.models import XFormInstanceSQL, CaseTransaction, LedgerTransaction
+from corehq.form_processor.submission_post import SubmissionPost
+from corehq.form_processor.utils.general import should_use_sql_backend
+from corehq.form_processor.utils.xform import _get_form
+from corehq.sql_db.util import get_db_alias_for_partitioned_doc
+from couchforms.models import XFormInstance
+
+
+def pre_process_unfinished_stub(stub):
+    if stub.saved:
+        # ignore for now
+        return
+
+    if not should_use_sql_backend(stub.domain):
+        # ignore for couch domains
+        stub.delete()
+        return
+
+    form_id = stub.xform_id
+    try:
+        form = FormAccessorSQL.get_form(form_id)
+    except XFormNotFound:
+        # form doesn't exist which means the failure probably happend during saving so
+        # let mobile handle re-submitting it
+        stub.delete()
+        return
+
+    _reprocess_form(form)
+
+
+def reprocess_xform_error(form):
+    """
+    Attempt to re-process an error form. This was created specifically to address
+    the issue of out of order forms and child cases (form creates child case before
+    parent case has been created).
+
+    See http://manage.dimagi.com/default.asp?250459
+    :param form_id: ID of the error form to process
+    """
+    if not form:
+        raise Exception('Form with ID {} not found'.format(form.form_id))
+
+    if not form.is_error:
+        raise Exception('Form was not an error form: {}={}'.format(form.form_id, form.doc_type))
+
+    return _reprocess_form(form)
+
+
+def reprocess_xform_error_by_id(form_id, domain=None):
+    form = _get_form(form_id)
+    if domain and form.domain != domain:
+        raise Exception('Form not found')
+    return reprocess_xform_error(form)
+
+
+def _reprocess_form(form):
+    # reset form state prior to processing
+    if should_use_sql_backend(form.domain):
+        form.state = XFormInstanceSQL.NORMAL
+    else:
+        form.doc_type = 'XFormInstance'
+
+    form.initial_processing_complete = True
+    form.problem = None
+
+    cache = FormProcessorInterface(form.domain).casedb_cache(
+        domain=form.domain, lock=True, deleted_ok=True, xforms=[form]
+    )
+    with cache as casedb:
+        case_stock_result = SubmissionPost.process_xforms_for_cases([form], casedb)
+
+        if case_stock_result:
+            stock_result = case_stock_result.stock_result
+            if stock_result:
+                assert stock_result.populated
+
+            cases = case_stock_result.case_models
+            if should_use_sql_backend(form.domain):
+                cases = _filter_already_processed_cases(form, cases)
+                for case in cases:
+                    CaseAccessorSQL.save_case(case)
+
+                ledgers = _filter_already_processed_ledgers(form, stock_result.models_to_save)
```

Comment:
thanks, not sure why those checks were there since it's never None at this stage: 1052425